core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockStore.java

```diff
@@ -0,0 +1,307 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block;
+
+import alluxio.exception.BlockAlreadyExistsException;
+import alluxio.exception.BlockDoesNotExistException;
+import alluxio.exception.ExceptionMessage;
+import alluxio.exception.UfsBlockAccessTokenUnavailableException;
+import alluxio.worker.block.io.BlockReader;
+import alluxio.worker.block.meta.UnderFileSystemBlockMeta;
+
+import com.google.common.base.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * This class manages the virtual blocks in the UFS for delegated UFS reads/writes.
+ *
+ * The usage pattern:
+ *  acquireAccess(blockMeta, maxConcurrency)
+ *  cleanup(sessionId, blockId)
+ *  releaseAccess(sessionId, blockId)
+ *
+ * If the client is lost before releasing or cleaning up the session, the session cleaner will
+ * clean the data.
+ */
+public final class UnderFileSystemBlockStore {
+  private static final Logger LOG = LoggerFactory.getLogger(UnderFileSystemBlockStore.class);
+
+  private final ReentrantLock mLock = new ReentrantLock();
```

Comment:
I personally prefer to using lock directly as that avoids creating a new object in java which is not free as java doesn't keep objects in stack.